src/library_pthread.js

```diff
@@ -892,24 +872,39 @@ var LibraryPThread = {
 #endif
 
     for (;;) {
-      var detach_state = Atomics.load(HEAPU32, (thread + {{{ C_STRUCTS.pthread.detach_state }}} ) >> 2);
-      if (detach_state == {{{ cDefine('DT_EXITING') }}}) { // Exiting?
+      // The thread we are joining with must be either DT_JOINABLE or
+      // DT_EXITING.  If its DT_EXITING then we move it to DT_EXITED and
+      // we are done.   If its DT_JOINABLE we keep waiting.
+      var old_state = Atomics.compareExchange(HEAP32,
+        (thread + {{{ C_STRUCTS.pthread.detach_state }}}) >> 2,
+        {{{ cDefine('DT_EXITING') }}},
+        {{{ cDefine('DT_EXITED') }}}
+      );
+      if (old_state == {{{ cDefine('DT_EXITING') }}}) {
+#if PTHREADS_DEBUG
+        err('thread 0x' + thread.toString(16) + ' successfully joined');
+#endif
+        // We successfully marked the tread as DT_EXITED
         if (status) {
           var result = Atomics.load(HEAPU32, (thread + {{{ C_STRUCTS.pthread.result }}} ) >> 2);
           {{{ makeSetValue('status', 0, 'result', 'i32') }}};
         }
-        // Mark the thread as exited.
-        PThread.setDetachState(thread, {{{ cDefine('DT_EXITED') }}});
         if (!ENVIRONMENT_IS_PTHREAD) cleanupThread(thread);
         else postMessage({ 'cmd': 'cleanupThread', 'thread': thread });
         return 0;
       }
+#if ASSERTIONS
+      assert(old_state === {{{ cDefine('DT_JOINABLE') }}}, 'pthread_join attempted on thread 0x' + thread.toString(16) + ', which is in an invalid state:' + old_state);
+#else
+      if (old_state !== {{{ cDefine('DT_JOINABLE') }}}) return {{{ cDefine('EINVAL') }}};
+#endif
```

Comment:
Perhaps we should print an error, instead of `abort()`-ing when `ASSERTIONS` are enabled?

```suggestion
      if (old_state !== {{{ cDefine('DT_JOINABLE') }}}) {
#if ASSERTIONS
        err('pthread_join attempted on thread 0x' + thread.toString(16) + ', which is in an invalid state:' + old_state);
#endif
        return {{{ cDefine('EINVAL') }}};
      }

Reply:
I considered that yes.  But then I looked at the existing musl code it just aborts in this case.  I think it means your program is invalid if this happens:

See:

https://github.com/emscripten-core/emscripten/blob/d90468c032e90a142f1bc4fa0f1f44467fd7c825/system/lib/libc/musl/src/thread/pthread_join.c#L17

Reply:
Ah yes, it is considered to be undefined behavior. I currently patched that with this:
https://github.com/emscripten-core/emscripten/blob/3d39427072954fea76a2ce405a55d859f74f3925/system/lib/libc/musl/src/thread/pthread_join.c#L31-L40
in PR #13007, since the POSIX test suite depend on this (`__pthread_timedjoin_np` is also used by `pthread_detach`, so detaching an already-detached thread would then also crash).

Reply:
Is it enough to check for this on entry to pthread_join?  Its seems like that is enough to make the testsuite happy anyway.

Reply:
I have a followup patch that looked very much like https://github.com/emscripten-core/emscripten/blob/3d39427072954fea76a2ce405a55d859f74f3925/system/lib/libc/musl/src/thread/pthread_join.c#L31-L40

Sorry my work seems to be overlapping so with things you have already done :-/

Reply:
> Is it enough to check for this on entry to pthread_join?

This assertion cannot hurt I think.

> Its seems like that is enough to make the testsuite happy anyway.

Ah interesting. Perhaps it was already caught by this check?:
https://github.com/emscripten-core/emscripten/blob/f065db68f698469ff5db8415d872b4e2d223c396/system/lib/pthread/pthread_join.c#L23-L26

> Sorry my work seems to be overlapping so with things you have already done :-/

No worries. I'm still a student, so it's a great learning experience anyway.